Add a -public-linkage-for-metadata-accessors IRGen flag for LLDB.

adrian-prantl opened the pull request:
In LLDB expressions, references to private metadata accessors may be
emitted and need to be bound to symbols available in the attached
program, even if these symbols are only supposed to have private
visibility within the program.

Also rdar://problem/48018240

adrian-prantl commented:
@swift-ci test

slavapestov reviewed (commented):
This feels a bit arbitrary. What about all the other accessor functions IRGen emits, or even just private methods? Maybe instead do a post-processing pass on the LLVM module to make everything public? I also thought at some point the lldb expression evaluator walked to AST making everything public - is that no longer happening?

I’m also worried that “plumbing” boolean parameters like this through multiple layers of calls will lead to a proliferation of boolean conditions that will become hard to reason about as these sort of flags accumulate. I’d be happy to discuss other ways of solving the underlying problem.

rjmccall commented:
> This feels a bit arbitrary. What about all the other accessor functions IRGen emits, or even just private methods? Maybe instead do a post-processing pass on the LLVM module to make everything public?

Or have IRGen's linkage computation understand that private things from the execution context aren't actually object-file-private when running in the debugger.  We do something similar with private declarations when building for LTO.

adrian-prantl commented:
> > This feels a bit arbitrary. What about all the other accessor functions IRGen emits, or even just private methods? Maybe instead do a post-processing pass on the LLVM module to make everything public?
> 
> Or have IRGen's linkage computation understand that private things from the execution context aren't actually object-file-private when running in the debugger. We do something similar with private declarations when building for LTO.

Can you elaborate on this a bit? So, still have an IRGen flag that specifies that this is for the expression evaluator, but have it apply to more kinds of symbols? What's a reasonable boundary?

A quick grep for LTO in IRGen didn't find any results, what should I be looking for?

rjmccall commented:
Sorry, I meant for WMO.  See `getIRLinkage` in GenDecl.cpp.

adrian-prantl committed:
Add a -force-public-linkage IRGen flag for LLDB.

In LLDB expressions, references to private metadata accessors may be
emitted and need to be bound to symbols available in the attached
program, even if these symbols are only supposed to have private
visibility within the program.

Also rdar://problem/48018240

adrian-prantl commented:
Sorry for the delay.

@rjmccall Is this closer to what you had in mind? The option to communicate that we are in the LLDB expression evaluator is still there, but it is now localized to `getIRLinkage()`.

adrian-prantl commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -1462,6 +1462,9 @@ getIRLinkage(const UniversalLinkageInfo &info, SILLinkage linkage,
   case SILLinkage::Private: {
+    if (info.forcePublicDecls())
+      return getIRLinkage(info, SILLinkage::SharedExternal, isDefinition,
+                          isWeakImported);
```
I'm also not sure whether this is the best possible set of flags. It appears to work...

rjmccall replied:
It probably ought to be based on whether this is a definition, so that definitions are shared (or maybe even private) and non-definitions are public-external.

slavapestov commented:
@adrian-prantl LLDB defines a method named `SwiftASTManipulator::MakeDeclarationsPublic()`. Is it not getting called anymore? Or its not doing the right thing?

rjmccall commented on include/swift/IRGen/Linking.h:
```diff
@@ -49,12 +49,15 @@ class UniversalLinkageInfo {
   bool HasMultipleIGMs;
 
+  /// Used by the LLDB expression evaluator.
+  bool ForcePublicDecls;
```
The comment here should explain what this means, not how it's used.

rjmccall reviewed (commented):
Yeah, that's more what I was thinking, thanks.

adrian-prantl commented:
> @adrian-prantl LLDB defines a method named `SwiftASTManipulator::MakeDeclarationsPublic()`. Is it not getting called anymore? Or its not doing the right thing?

IIUC the problem I'm trying to solve here is that Swift metadata accessors particularly are transparent to the AST and only injected/created by IRGen.

slavapestov commented:
@adrian-prantl The linkage of a metadata accessor is computed from the access level of the nominal type. So if the type is public the accessor should be public too. Is this not happening in some cases?

adrian-prantl commented:
> @adrian-prantl The linkage of a metadata accessor is computed from the access level of the nominal type. So if the type is public the accessor should be public too. Is this not happening in some cases?

The type in the example was private.

```
For context, the expression is printing a global variable

private struct FixedContainer {
  var s = S() // S is a resilient struct and defined in a different module
}
private var g_main_c = FixedContainer()

```

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

slavapestov commented:
> The type in the example was private.

Right, but doesn't `SwiftASTManipulator::MakeDeclarationsPublic()` make everything public? That's the part I'm confused about.

adrian-prantl commented:
As far as I understand the type in question is not part of the AST walked by SwiftASTManipulator::Publicist. It walks the expression source file, but our type is part of the swiftmodule that is being debugged.

adrian-prantl commented:
Note that just because I merged this, doesn't mean we shouldn't still replace this with something better if we find it, but this is blocking all the resilience support.

slavapestov commented:
@adrian-prantl No worries.

If the Decl is part of the swiftmodule being debugged, why is lldb emitting the type metadata accessor?

jrose-apple commented:
It's emitting a reference to the existing accessor (a declaration rather than a definition).

slavapestov commented:
@jrose-apple I see. Perhaps it makes more sense to elevate private linkage to internal when emitting code with -g?
